src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -23,50 +26,153 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             }
 
-            var semanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
+            var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
+            var cSharpSemanticRanges = await GetCSharpSemanticRanges(codeDocument, textDocumentIdentifier, range: null, cancellationToken);
+
+            var combinedSemanticRanged = CombineSemanticRanges(razorSemanticRanges, cSharpSemanticRanges);
 
             IReadOnlyList<int> previousResults = null;
 
             if (previousResultId != null)
             {
                 _semanticTokensCache.TryGetValue(previousResultId, out previousResults);
             }
-            var newTokens = ConvertSemanticRangesToSemanticTokens(semanticRanges, codeDocument);
+            var newTokens = ConvertSemanticRangesToSemanticTokens(combinedSemanticRanged, codeDocument);
 
             if (previousResults is null)
             {
                 return newTokens;
             }
 
-            var semanticEdits = SemanticTokensEditsDiffer.ComputeSemanticTokensEdits(newTokens, previousResults);
+            var razorSemanticEdits = SemanticTokensEditsDiffer.ComputeSemanticTokensEdits(newTokens, previousResults);
+
+            return razorSemanticEdits;
+        }
+
+        private IReadOnlyList<SemanticRange> CombineSemanticRanges(params IReadOnlyList<SemanticRange>[] rangesArray)
+        {
+            var newList = new List<SemanticRange>();
+            foreach(var list in rangesArray)
+            {
+                if(list != null)
+                {
+                    newList.AddRange(list);
+                }
+            }
+
+            newList.Sort();
+
+            return newList;
+        }
+
+        private async Task<IReadOnlyList<SemanticRange>> GetCSharpSemanticRanges(
+            RazorCodeDocument codeDocument,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range range,
+            CancellationToken cancellationToken)
+        {
+            var parameter = new SemanticTokensParams
+            {
+                TextDocument = textDocumentIdentifier,
+            };
+
+            var request = _languageServer.SendRequest(LanguageServerConstants.RazorProvideSemanticTokensEndpoint, parameter);
+            var cSharpResponses = await request.Returning<SemanticTokens>(cancellationToken);
+            if(cSharpResponses is null)
+            {
+                return null;
+            }
+
+            SemanticRange previousSemanticRange = null;
+            var razorRanges = new List<SemanticRange>();
+            for(var i = 0; i < cSharpResponses.Data.Length; i+=5){
+                var lineDelta = cSharpResponses.Data[i];
+                var charDelta = cSharpResponses.Data[i+1];
+                var length = cSharpResponses.Data[i+2];
+                var tokenType = cSharpResponses.Data[i+3];
+                var tokenModifiers = cSharpResponses.Data[i+4];
```

Comment:
Wish there was some syntax sugar which could allow us to simplify this unpacking like we do with tuples.

Reply:
Yeah, it would be nice is this was less wordy, but it was the best way I could come up with to be explicit about the data we're using here.